exporter/awsemfexporter/emf_exporter.go

```diff
@@ -0,0 +1,194 @@
+		config:           config,
+		retryCnt:         *awsConfig.MaxRetries,
+		logger:           logger,
+	}
+	emfExporter.groupStreamToPusherMap = map[string]map[string]Pusher{}
+
+	return emfExporter, nil
+}
+
+func (emf *emfExporter) pushMetricsData(_ context.Context, md pdata.Metrics) (droppedTimeSeries int, err error) {
+	expConfig := emf.config.(*Config)
+	logGroup := "/metrics/default"
+	logStream := "otel-stream"
+	// override log group if customer has specified Resource Attributes service.name or service.namespace
+	putLogEvents, totalDroppedMetrics, namespace := generateLogEventFromMetric(md)
+	if namespace != "" {
+		logGroup = fmt.Sprintf("/metrics/%s", namespace)
+	}
+	// override log group if found it in exp configuration, this configuration has top priority. However, in this case, customer won't have correlation experience
+
+	if len(expConfig.LogGroupName) > 0 {
+		logGroup = expConfig.LogGroupName
+	}
+	if len(expConfig.LogStreamName) > 0 {
+		logStream = expConfig.LogStreamName
+	}
+	pusher := emf.getPusher(logGroup, logStream)
+	if pusher != nil {
+		for _, ple := range putLogEvents {
+			returnError := pusher.AddLogEntry(ple)
+			if returnError != nil && !emf.config.(*Config).LocalMode {
+				err = wrapErrorIfBadRequest(&returnError)
+			}
+			if err != nil {
+				return totalDroppedMetrics, err
+			}
+		}
+		returnError := pusher.ForceFlush()
+		if returnError != nil && !emf.config.(*Config).LocalMode {
+			err = wrapErrorIfBadRequest(&returnError)
+		}
+		if err != nil {
+			return totalDroppedMetrics, err
+		}
+	}
+	return totalDroppedMetrics, nil
+}
+
+func (emf *emfExporter) getPusher(logGroup, logStream string) Pusher {
+	emf.pusherMapLock.Lock()
+	defer emf.pusherMapLock.Unlock()
+
+	var ok bool
+	var streamToPusherMap map[string]Pusher
+	if streamToPusherMap, ok = emf.groupStreamToPusherMap[logGroup]; !ok {
+		streamToPusherMap = map[string]Pusher{}
+		emf.groupStreamToPusherMap[logGroup] = streamToPusherMap
+	}
+
+	var pusher Pusher
+	if pusher, ok = streamToPusherMap[logStream]; !ok {
+		pusher = NewPusher(aws.String(logGroup), aws.String(logStream), emf.retryCnt, emf.svcStructuredLog)
+		streamToPusherMap[logStream] = pusher
+	}
+	return pusher
+}
+
+func (emf *emfExporter) ConsumeMetrics(ctx context.Context, md pdata.Metrics) error {
+	exporterCtx := obsreport.ExporterContext(ctx, "emf.exporterFullName")
+
+	_, err := emf.pushMetricsData(exporterCtx, md)
+	return err
+}
+
+// Shutdown stops the exporter and is invoked during shutdown.
+func (emf *emfExporter) Shutdown(ctx context.Context) error {
+	emf.pusherMapLock.Lock()
+	defer emf.pusherMapLock.Unlock()
+
+	var err error
+	for _, streamToPusherMap := range emf.groupStreamToPusherMap {
+		for _, pusher := range streamToPusherMap {
+			if pusher != nil {
+				returnError := pusher.ForceFlush()
+				if returnError != nil && !emf.config.(*Config).LocalMode {
+					err = wrapErrorIfBadRequest(&returnError)
+				}
+				if err != nil {
```

Comment:
I think we should log but not return err here, mainly to allow going through all the pushers